src/lib/ProjectM36/Base.hs

```diff
@@ -527,3 +538,75 @@ instance Hashable DatabaseContextFunction where
                            
 instance Eq DatabaseContextFunction where                           
   f1 == f2 = dbcFuncName f1 == dbcFuncName f2 
+
+
+-- for creating arbitrary tuples 
+instance Arbitrary Text where
+  arbitrary = pack <$> elements ["Mary", "Johnny", "Sunny", "Ted"]
+
+instance Arbitrary Day where
+  arbitrary = ModifiedJulianDay <$> (arbitrary :: Gen Integer)
+
+instance Arbitrary Attribute where
+  arbitrary = genericArbitrary
+
+instance (Arbitrary a) => Arbitrary (V.Vector a) where
+    arbitrary = do
+      maxbound <- choose (0,5)
+      V.fromList <$> vectorOf maxbound arbitrary
+
+--not sure what's use on this one
+instance {-# OVERLAPPING #-} Arbitrary TypeVarMap where
+    arbitrary = undefined
+
+instance Arbitrary UTCTime where
+ arbitrary = UTCTime <$> arbitrary <*> (secondsToDiffTime <$> arbitrary)
+
+instance Arbitrary B.ByteString where
+  arbitrary = B.pack <$> (arbitrary :: Gen String)
+
+defAttributeNames = ["Number","People","Time","Thing"]
+defAtomtypes = [IntAtomType, TextAtomType, DayAtomType, TextAtomType]
+defAttributes = V.fromList $ zipWith Attribute defAttributeNames defAtomtypes
+
+instance Arbitrary RelationTuple where
+  arbitrary = do
+    a <- IntAtom <$> (arbitrary :: Gen Int)
+    b <- TextAtom <$> (arbitrary :: Gen Text)
+    c <- DayAtom <$> (arbitrary :: Gen Day)
+    d <- TextAtom <$> (arbitrary :: Gen Text)
+    let sample = V.fromList [a,b,c,d]
+    return $ RelationTuple defAttributes sample
+
+instance Arbitrary Relation where
+  arbitrary = do
+    list <- listOf (arbitrary :: Gen RelationTuple)
+    return $ Relation defAttributes (RelationTupleSet list)
+
+arbitraryRelationTuple attris = do
+  let attrisG = V.fromList <$>  mapM (arbitrary'.getAtomType) (V.toList attris)
+  RelationTuple attris <$> attrisG
+
+arbitraryRel attris bound@(min,max) = do
+  num <- choose (min,max)
+  let tupleListG = vectorOf num (arbitraryRelationTuple attris)
+  Relation attris <$> (RelationTupleSet <$> tupleListG)
+
+instance Arbitrary AtomType where
+  arbitrary = genericArbitrary
+
+instance Arbitrary Atom where
+  arbitrary = genericArbitrary
+
+instance ToADTArbitrary Atom
+
+arbitrary' :: AtomType -> Gen Atom
+arbitrary' IntAtomType  = IntAtom <$> (arbitrary :: Gen Int)
+arbitrary' IntegerAtomType  = IntegerAtom <$> (arbitrary :: Gen Integer)
+arbitrary' TextAtomType = TextAtom <$> (arbitrary :: Gen Text)
+arbitrary' DayAtomType  = DayAtom <$> (arbitrary :: Gen Day)
+arbitrary' BoolAtomType  = BoolAtom <$> (arbitrary :: Gen Bool)
+arbitrary' _            = undefined
```

Comment:
Yes, I implement simple atomtypes for a start, only to get some feedback. I have trouble with imaging the usage of other atomtypes. I don't know if arbitrary Constructed Atomtypes is needed since it is user-defined. When I random things, it feels I didn't offer enough meaningful data. But I guess this arbitrary feature is for testing some operation functionality, right? Sub-relation is a good point, thanks for reminding me.

I'll look into sub-relations and construct atomtypes, making it complete.
and do the cleanup you mentioned. 

Thanks again for the comment and review.

Reply:
arbitrary IntAtom is added now. 

However, I noticed,
TutorialD (master/main): x:=relation{name Text, age Int}{tuple{name "Steve", age 40},tuple{name "Mike", age 31}}
ERR: AtomTypeMismatchError IntAtomType IntegerAtomType

And there is no intP for atomP in code. So there seems no input way for IntAtom.

reference: issue [#138](https://github.com/agentm/project-m36/issues/138)

Reply:
I think that a version one of this patch need not necessarily include arbitrary attributes. But, if you think it's easy-to-add, that's fine.

Regarding the `Int` mismatch, the issue is that the default type for `40` is Integer. Project:M36 doesn't support typeclasses yet, but it would be nice to infer the type from the attributes. I'll open a ticket for this specific issue.

Reply:
I added an `int()` function in #201.

Example: `:showexpr relation{a Int}{tuple{a int(4)}}`

Reply:
I was trying to create an arbitrary relation with sub-relation. So first, I need make an attribute list with an relation attribute in it.

TutorialD (master/main): a :: {name Text, age Integer}
TutorialD (master/main): createarbitraryrelation b {name Text, age Integer, rel a} 1-2
TutorialD (master/main): :showexpr b
tutd: Prelude.undefined
It is becaue I haven't defined `TypeVariableType`
`arbitrary' (TypeVariableType tvname) = undefined`

How could I define an relation with sub-relation?

Reply:
To make arbitrary sub-relation occur, then a way to define its attributes in `DEFINE ::` or `ASSIGN :=` is need.
Something like `a :: {name Text, rel relation{name Text, age Integer} }` is needed.
I haven't figure out the whole expression yet.

And I don't understand these two things, ConstructedAtomType and TypeVariable, for now. So I don't know when an arbitrary instance of these two will be used. 

I don't think arbitrary attributes is necessary either.

Reply:
I notice that an arbitrary subrel can be created by `group` after `createarbitraryrelation`. 
```
createarbitraryrelation s {city Text,s# Integer,sname Text,status Integer} 1-2
showexpr s group ({sname,status,s#} as subrel)
```

So I guess the implementation is unncessary.

Reply:
The lack of ability to construct to create a relation type is just a bug. I'll work on fixing that in master so that you can proceed. Thanks for the report!

Reply:
Thanks! What a great project!